main/src/cgeo/geocaching/ui/CalculateButton.java

```diff
@@ -0,0 +1,265 @@
+
+    // The buttons own state is stored in a 'state' object to facilitate easy saving and restoring.
+    private ButtonData buttonData;
+
+    // Data used to capture the state of this particular button such that it can be restored again later.
+    public static class ButtonData implements Serializable {
+        ValueType type = ValueType.INPUT_VAL;
+        char inputVal;                  // Value obtained from CoordinateInputDialog.
+        char autoChar = EMPTY_CHAR;     // Character obtained by automatically 'counting up' variable names.
+        char customChar;                // User defined character.
+
+        ButtonData() { }
+
+        public ButtonData(final JSONObject json) {
+            type = ValueType.values()[json.optInt("type", 0)];
+            inputVal   = (char) json.optInt("inputVal",   ERROR_CHAR);
+            autoChar   = (char) json.optInt("autoChar",   ERROR_CHAR);
+            customChar = (char) json.optInt("customChar", ERROR_CHAR);
+        }
+
+        public JSONObject toJSON() throws JSONException {
+            final JSONObject returnValue = new JSONObject();
+
+            returnValue.put("type", type.ordinal());
+            returnValue.put("inputVal", inputVal);
+            returnValue.put("autoChar", autoChar);
+            returnValue.put("customChar", customChar);
+
+            return returnValue;
+        }
+    }
+
+    private class CoordDigitClickListener implements View.OnClickListener {
+        @Override
+        public void onClick(final View view) {
+            toggleType();
+        }
+    }
+
+    public CalculateButton(final Context context) {
+        super(context);
+        setup();
+    }
+
+    public CalculateButton(final Context context, final AttributeSet attrs) {
+        super(context, attrs);
+        setup();
+    }
+
+    private void setup() {
+        setData(new ButtonData());
+        butt.setOnClickListener(new CoordDigitClickListener());
+    }
+
+    // The 'Label' is the 'name' that is to be displayed on the button
+    public char getLabel() {
+        final char returnValue;
+        switch (getType()) {
+            case INPUT_VAL: returnValue = buttonData.inputVal;   break;
+            case AUTO_CHAR: returnValue = buttonData.autoChar;   break;
+            case CUSTOM:    returnValue = buttonData.customChar; break;
+            default:        returnValue = '*'; // Should never happen.
+        }
+
+        return returnValue;
+    }
+
+    public void setLabel(final char lable) {
+        butt.setText(String.valueOf(lable));
+    }
+
+    public ValueType getType() {
+        return buttonData.type;
+    }
+
+    public void setType(final ValueType type) {
+        buttonData.type = type;
+        setBackgroundColour();
+    }
+
+    // Buttons displaying a custom value are given a 'steel-like' appearance so as to distinguish them from regular buttons.
+    private  void setBackgroundColour() {
+        if (buttonData.type == ValueType.CUSTOM) {
```

Comment:
I ended up changing this to a switch rather than overriding methods in the enum as having both the 'INPUT_CHAR' and 'AUTO_CHAR' cases run the same code was a little clunky.  As a switch though it's quite neat :)